File: src/main/java/com/trade/securities/adapter/in/web/StockQuotationController.java
package com.trade.securities.adapter.in.web;

import com.trade.securities.application.port.in.GetFscBeneficiaryCertificateUseCase;
import com.trade.securities.application.port.in.GetFscNewShareCertificateUseCase;
import com.trade.securities.application.port.in.GetFscStockPriceUseCase;
import com.trade.securities.application.port.in.GetFscStockSubscriptionRightUseCase;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

@RestController
@RequestMapping("/api/stock-quotation")
@RequiredArgsConstructor
public class StockQuotationController {

    private final GetFscStockPriceUseCase getFscStockPriceUseCase;
    private final GetFscNewShareCertificateUseCase getFscNewShareCertificateUseCase;
    private final GetFscBeneficiaryCertificateUseCase getFscBeneficiaryCertificateUseCase;
    private final GetFscStockSubscriptionRightUseCase getFscStockSubscriptionRightUseCase;

    @GetMapping("/price")
    public Map<String, Object> getStockPrices(
            @RequestParam(defaultValue = "1") int pageNo,
            @RequestParam(defaultValue = "10") int numOfRows,
            @RequestParam String basDt,
            @RequestParam(required = false) String itmsNm,
            @RequestParam(required = false) String likeItmsNm) {
        return getFscStockPriceUseCase.getStockPrices(pageNo, numOfRows, basDt, itmsNm, likeItmsNm);
    }

    @GetMapping("/new-share")
    public Map<String, Object> getNewShareCertificates(
            @RequestParam(defaultValue = "1") int pageNo,
            @RequestParam(defaultValue = "10") int numOfRows,
            @RequestParam String basDt,
            @RequestParam(required = false) String itmsNm,
            @RequestParam(required = false) String likeItmsNm) {
        return getFscNewShareCertificateUseCase.getNewShareCertificates(pageNo, numOfRows, basDt, itmsNm, likeItmsNm);
    }

    @GetMapping("/beneficiary")
    public Map<String, Object> getBeneficiaryCertificates(
            @RequestParam(defaultValue = "1") int pageNo,
            @RequestParam(defaultValue = "10") int numOfRows,
            @RequestParam String basDt,
            @RequestParam(required = false) String itmsNm,
            @RequestParam(required = false) String likeItmsNm) {
        return getFscBeneficiaryCertificateUseCase.getBeneficiaryCertificates(pageNo, numOfRows, basDt, itmsNm,
                likeItmsNm);
    }

    @GetMapping("/subscription-right")
    public Map<String, Object> getSubscriptionRights(
            @RequestParam(defaultValue = "1") int pageNo,
            @RequestParam(defaultValue = "10") int numOfRows,
            @RequestParam String basDt,
            @RequestParam(required = false) String itmsNm,
            @RequestParam(required = false) String likeItmsNm) {
        return getFscStockSubscriptionRightUseCase.getSubscriptionRights(pageNo, numOfRows, basDt, itmsNm, likeItmsNm);
    }
}

File: src/main/kotlin/com/trade/securities/adapter/in/web/StockQuotationController.kt
package com.trade.securities.adapter.`in`.web

import com.trade.securities.application.port.`in`.GetFscBeneficiaryCertificateUseCase
import com.trade.securities.application.port.`in`.GetFscNewShareCertificateUseCase
import com.trade.securities.application.port.`in`.GetFscStockPriceUseCase
import com.trade.securities.application.port.`in`.GetFscStockSubscriptionRightUseCase
import jakarta.enterprise.context.ApplicationScoped
import jakarta.ws.rs.DefaultValue
import jakarta.ws.rs.GET
import jakarta.ws.rs.Path
import jakarta.ws.rs.Produces
import jakarta.ws.rs.QueryParam
import jakarta.ws.rs.core.MediaType

@ApplicationScoped
@Path("/api/stock-quotation")
@Produces(MediaType.APPLICATION_JSON)
class StockQuotationController(
    private val getFscStockPriceUseCase: GetFscStockPriceUseCase,
    private val getFscNewShareCertificateUseCase: GetFscNewShareCertificateUseCase,
    private val getFscBeneficiaryCertificateUseCase: GetFscBeneficiaryCertificateUseCase,
    private val getFscStockSubscriptionRightUseCase: GetFscStockSubscriptionRightUseCase
) {

    @GET
    @Path("/price")
    fun getStockPrices(
        @QueryParam("pageNo") @DefaultValue("1") pageNo: Int,
        @QueryParam("numOfRows") @DefaultValue("10") numOfRows: Int,
        @QueryParam("basDt") basDt: String,
        @QueryParam("itmsNm") itmsNm: String?,
        @QueryParam("likeItmsNm") likeItmsNm: String?
    ): Map<String, Any?> {
        return getFscStockPriceUseCase.getStockPrices(pageNo, numOfRows, basDt, itmsNm, likeItmsNm)
    }

    @GET
    @Path("/new-share")
    fun getNewShareCertificates(
        @QueryParam("pageNo") @DefaultValue("1") pageNo: Int,
        @QueryParam("numOfRows") @DefaultValue("10") numOfRows: Int,
        @QueryParam("basDt") basDt: String,
        @QueryParam("itmsNm") itmsNm: String?,
        @QueryParam("likeItmsNm") likeItmsNm: String?
    ): Map<String, Any?> {
        return getFscNewShareCertificateUseCase.getNewShareCertificates(pageNo, numOfRows, basDt, itmsNm, likeItmsNm)
    }

    @GET
    @Path("/beneficiary")
    fun getBeneficiaryCertificates(
        @QueryParam("pageNo") @DefaultValue("1") pageNo: Int,
        @QueryParam("numOfRows") @DefaultValue("10") numOfRows: Int,
        @QueryParam("basDt") basDt: String,
        @QueryParam("itmsNm") itmsNm: String?,
        @QueryParam("likeItmsNm") likeItmsNm: String?
    ): Map<String, Any?> {
        return getFscBeneficiaryCertificateUseCase.getBeneficiaryCertificates(pageNo, numOfRows, basDt, itmsNm, likeItmsNm)
    }

    @GET
    @Path("/subscription-right")
    fun getSubscriptionRights(
        @QueryParam("pageNo") @DefaultValue("1") pageNo: Int,
        @QueryParam("numOfRows") @DefaultValue("10") numOfRows: Int,
        @QueryParam("basDt") basDt: String,
        @QueryParam("itmsNm") itmsNm: String?,
        @QueryParam("likeItmsNm") likeItmsNm: String?
    ): Map<String, Any?> {
        return getFscStockSubscriptionRightUseCase.getSubscriptionRights(pageNo, numOfRows, basDt, itmsNm, likeItmsNm)
    }
}
